### PR TITLE
Check for and log information about stale osquery database lock files

### PR DESCRIPTION
**Pull request by RebeccaMahany**
Relates to https://github.com/kolide/launcher/issues/2004.

launcher is currently completely unable to recover if the osquery lock file is left behind. However, we want to be cautious about removing this file. For now, we detect whether there's a stale lock file, and collect + log information about it if so.

Also adds a test for launching an instance, plus a test for detecting lock file presence.

### CODE REVIEW

File: pkg/osquery/runtime/osqueryinstance.go [hunk 1]
```diff
@@ -249,6 +249,12 @@ func (i *OsqueryInstance) Launch() error {
 		return fmt.Errorf("could not calculate osquery file paths: %w", err)
 	}
 
+	// Make sure no lock files exist that could prevent osquery from starting up
+	if err := i.cleanUpOldDatabaseLock(ctx, paths); err != nil {
```
Comment (James-Pickett) [inline pkg/osquery/runtime/osqueryinstance.go:253]: probably want to do this in a loop, windows can be so finicky about deleting files in use

**Review by James-Pickett — DISMISSED** [review]

File: pkg/osquery/runtime/osqueryinstance.go (hunk 1 above)
Comment (RebeccaMahany) [reply pkg/osquery/runtime/osqueryinstance.go:253]: Oh yes, thank you! I'll move this to a backoff/retry

**Review by James-Pickett — DISMISSED** [review]

**Review by zackattack01 — APPROVED** [review]

File: pkg/osquery/runtime/osqueryinstance.go [hunk 2]
```diff
@@ -702,6 +708,45 @@ func calculateOsqueryPaths(rootDirectory string, registrationId string, runId st
...
+func (i *OsqueryInstance) cleanUpOldDatabaseLock(ctx context.Context, paths *osqueryFilePaths) error {
+	lockFilePath := filepath.Join(paths.databasePath, "LOCK")
+
+	lockFileInfo, err := os.Stat(lockFilePath)
+	if os.IsNotExist(err) {
+		// No lock file, nothing to do here
+		return nil
+	}
+
+	// We do a couple retries here -- Windows complains about removing files in use sometimes.
```
Comment (directionless) [inline pkg/osquery/runtime/osqueryinstance.go:723]: My hunch is that this is too aggressive -- if windows is blocking you from removing the file. the osquery probably has it open. I wonder if there's a way we can detect whether osquery is running, or what the state here really is. 

I think forcing the file out, with osquery running will lead to corrupt databases

**Review by directionless — CHANGES_REQUESTED** [review]

File: pkg/osquery/runtime/osqueryinstance.go (hunk 2 above)
Comment (RebeccaMahany) [reply pkg/osquery/runtime/osqueryinstance.go:723]: I'd had another idea for handling this --

> The osquery instance and osquery runner both currently have no visibility into the osquery logs -- they're processed by the log adapter, which is entirely separate. However, here and previously (when trying to figure out how to handle launcher falling back to a old version of osquery that is incompatible with the database -- some details https://github.com/kolide/launcher/issues/1937) we've tentatively wanted the osquery instance to be able to respond to particular logs from osquery. Do we want to move log processing into the osquery instance? Or open up some line of communication between the instance and the log adapter?

We could do something like:

- detect via logs that the lock file is in use (like we do when the pidfile is in use)
- gather information about what's using the lock file (like we do when the pidfile is in use)
- as long as it's a dead process, remove the lock file

What do you think?

Comment (RebeccaMahany) [reply pkg/osquery/runtime/osqueryinstance.go:723]: Actually, I could probably update this PR to do the second and third bullet points as-is, without having to mess around w the log adapter 🤔

File: pkg/osquery/runtime/osqueryinstance.go [hunk 3]
```diff
@@ -702,6 +712,79 @@
...
+		processStrs[i] = processStr(ctx, p)
+	}
+	infoToLog = append(infoToLog, "processes", processStrs)
+
+	return true, nil
+}
+
+func processStr(ctx context.Context, p *process.Process) string {
+	name := "unknown"
+	username := "unknown"
```
[the remark this answers is not in the document]
Comment (RebeccaMahany) [reply pkg/osquery/runtime/osqueryinstance.go:764]: These are not hardcoded credentials; this is a default display value that we log if we cannot determine the username that owns a given process. This value is only used for logging purposes.

**Review by James-Pickett — DISMISSED** [review]

**Review by James-Pickett — DISMISSED** [review]

File: pkg/osquery/runtime/runtime_helpers_windows.go [hunk 4]
```diff
@@ -80,3 +82,39 @@ func isExitOk(err error) bool {
 	}
 	return false
 }
+
+func getProcessesHoldingFile(ctx context.Context, pathToFile string) ([]*process.Process, error) {
```
Comment (directionless) [inline pkg/osquery/runtime/runtime_helpers_windows.go:86]: I think this is a fairly expensive call. Might be okay, given how infrequent it occurs. But we should know that it's not cheap

Comment (RebeccaMahany) [reply pkg/osquery/runtime/runtime_helpers_windows.go:86]: It takes about 6 seconds in test, which is definitely not nothing. But we'll only hit this if the lock file exists -- and in that case, we'll never successfully start up osquery anyway, so the delay is a moot point. I figure we can remove this once we have more info from the logs to figure out how to detect/remediate.

**Review by zackattack01 — DISMISSED** [review]

File: pkg/osquery/runtime/runtime_helpers.go [hunk 5]
```diff
@@ -33,3 +40,41 @@ func platformArgs() []string {
 func isExitOk(_ error) bool {
 	return false
 }
+
+func getProcessesHoldingFile(ctx context.Context, pathToFile string) ([]*process.Process, error) {
+	cmd, err := allowedcmd.Lsof(ctx, "-t", pathToFile)
```
Comment (directionless) [inline pkg/osquery/runtime/runtime_helpers.go:45]: I wish we had a better option, but I concur we do not.

File: pkg/osquery/runtime/runtime_helpers_windows.go (hunk 4 above)
Comment (directionless) [reply pkg/osquery/runtime/runtime_helpers_windows.go:86]: Yeah...

**Review by directionless — APPROVED** [review]

**Pull request merged**